src/CommonSocket.php

```diff
@@ -240,8 +262,149 @@ protected function createStream()
+        }
+
+        $null     = null;
+        $read     = [$this->getSocket()];
+        $readable = @stream_select($read, $null, $null, $this->recvTimeoutSec, $this->recvTimeoutUsec);
+        if ($readable > 0) {
+            $remainingBytes = $len;
+            $data           = $chunk = '';
+            while ($remainingBytes > 0) {
+                $chunk = fread($this->getSocket(), $remainingBytes);
+                if ($chunk === false) {
+                    $this->close();
+                    throw new \Kafka\Exception('Could not read ' . $len . ' bytes from stream (no data)');
+                }
+                if (strlen($chunk) === 0) {
+                    // Zero bytes because of EOF?
+                    if (feof($this->getSocket())) {
+                        $this->close();
+                        throw new \Kafka\Exception('Unexpected EOF while reading ' . $len . ' bytes from stream (no data)');
+                    }
+                    // Otherwise wait for bytes
+                    $readable = @stream_select($read, $null, $null, $this->recvTimeoutSec, $this->recvTimeoutUsec);
+                    if ($readable !== 1) {
+                        throw new \Kafka\Exception('Timed out reading socket while reading ' . $len . ' bytes with ' . $remainingBytes . ' bytes to go');
+                    }
+                    continue; // attempt another read
+                }
+                $data           .= $chunk;
+                $remainingBytes -= strlen($chunk);
+            }
+            if ($len === $remainingBytes || ($verifyExactLength && $len !== strlen($data))) {
+                // couldn't read anything at all OR reached EOF sooner than expected
+                $this->close();
+                throw new \Kafka\Exception('Read ' . strlen($data) . ' bytes instead of the requested ' . $len . ' bytes');
+            }
+
+            return $data;
+        }
+        if (false !== $readable) {
+            $res = stream_get_meta_data($this->getSocket());
+            if (! empty($res['timed_out'])) {
+                $this->close();
+                throw new \Kafka\Exception('Timed out reading ' . $len . ' bytes from stream');
+            }
+        }
+        $this->close();
+        throw new \Kafka\Exception('Could not read ' . $len . ' bytes from stream (not readable)');
+    }
+
+    // }}}
+    // {{{ public function selectWrite()
+
+    /**
+     * Write to the socket.
+     *
+     * @param string $buf The data to write
+     *
+     * @return integer
+     * @throws \Kafka\Exception
+     */
+    public function selectWrite($buf)
+    {
+        $null  = null;
+        $write = [$this->getSocket()];
+
+        // fwrite to a socket may be partial, so loop until we
+        // are done with the entire buffer
+        $failedWriteAttempts = 0;
+        $written             = 0;
+        $buflen              = strlen($buf);
+        while ($written < $buflen) {
```

Comment:
Same thing about complexity here, it's kind of complicated to understand everything here